### ISSUE TITLE
Downloading files with original names

### ISSUE THREAD

**Issue by linc** [vanilla/vanilla#3099]
When downloading an attachment, especially non-images, it would be a great enhancement to have the filenames corrected for the user. Currently, they get CDN gibberish because it's a direct link to the file. It does, however, show the correct name in the Editor thumbnails. Maybe we could do a selective pass-thru when attempting to download from those thumbnails so we can fix the filename?

**Comment by R-J** [comment]
**Basically** you can change the link in the attachments view so that it points to /download&name=$attachment['Name']&path=$attachment['Path']

Then you can add a function to editor plugin like that
~~~
public function rootController_download_create($sender, $args) {
    // Make sure it is not used to download "uploads/../conf/config.php" or worse.
    if (!stringBeginsWith(realpath(PATH_UPLOADS.DS.Gdn::request()->get('path'), PATH_UPLOADS.DS) {
        return;
    }
    // Set the name.
    header('Content-Disposition: attachment; filename='.Gdn::request()->get('name'));
    // Send file content.
    echo file_get_contents(PATH_UPLOADS.DS.Gdn::request()->get('path'));
}
~~~
I don't know if that realpath trick would be enough. I have only taken a quick look at the source, but I've seen something about the possibility to register different upload folders? If they are only subfolders of PATH_UPLOADS, the above code should work.

**Comment by linc** [comment]
Closed by #3409
